File: Love Sahaj/Java/editDistances.java
import java.util.HashMap;

public class editDistances {
    private class Pair {
        int first;
        int second;

        Pair(int f, int s) {
            this.first = f;
            this.second = s;
        }
    }

    private HashMap<Pair, Integer> dp = new HashMap<>();

    public int minDistance(String word1, String word2) {
        return minDistance(word1, word2, 0, 0, word1.length() - 1, word2.length() - 1);
    }

    private int minDistance(String word1, String word2, int i, int j, int len1, int len2) {
        if (i == len1 || j == len2) {
            Pair x = new Pair(i, j);
            if (dp.containsKey(x))
                return dp.get(x);

            int cost = Math.max(i, j);
            dp.put(x, cost);
            return cost;
        }

        if (word1.charAt(i) == word2.charAt(j)) {
            Pair x = new Pair(i, j);
            if (dp.containsKey(x))
                return dp.get(x);

            int cost = minDistance(word1, word2, i + 1, j + 1, len1, len2);
            dp.put(x, cost);
            return cost;
        }

        int cost = 1 + Math.min(minDistance(word1, word2, i + 1, j + 1, len1, len2), Math
                .min(minDistance(word1, word2, i, j + 1, len1, len2), minDistance(word1, word2, i + 1, j, len1, len2)));

        dp.put(new Pair(i, j), cost);
        return cost;
    }

    public static void main(String[] args) {
        editDistances sol = new editDistances();
        System.out.println(sol.minDistance("intention", "execution"));
    }
}

File: Love Sahaj/Python/editDistances.py
class Solution:
    def __init__(self):
        self.dp = dict()

    def minDistance(self, word1: str, word2: str) -> int:
        if len(word1) == 0 or len(word2) == 0:
            if (word1, word2) in self.dp:
                return self.dp[(word1, word2)]

            cost = max(len(word1), len(word2))
            self.dp[(word1, word2)] = cost
            return cost

        if word1[0] == word2[0] and (len(word1) == len(word2) == 1):
            if (word1, word2) in self.dp:
                return self.dp[(word1, word2)]

            cost = 0
            self.dp[(word1, word2)] = cost
            return cost

        if word1[0] != word2[0] and (len(word1) == len(word2) == 1):
            if (word1, word2) in self.dp:
                return self.dp[(word1, word2)]

            cost = 1
            self.dp[(word1, word2)] = cost
            return cost

        if word1[0] == word2[0]:
            if (word1, word2) in self.dp:
                return self.dp[(word1, word2)]

            cost = self.minDistance(word1[1:], word2[1:])
            self.dp[(word1, word2)] = cost
            return cost

        cost = 1 + min(self.minDistance(word1[1:], word2[1:]),
                       self.minDistance(word1, word2[1:]),
                       self.minDistance(word1[1:], word2))

        self.dp[(word1, word2)] = cost
        return cost


if __name__ == "__main__":
    sol = Solution()
    print(sol.minDistance(word1="horse", word2="ros"))
